assets/js/interactivity/store.ts

```diff
@@ -0,0 +1,240 @@
+				deepMerge( target[ key ], source[ key ] );
+			} else {
+				Object.assign( target, { [ key ]: source[ key ] } );
+			}
+		}
+	}
+};
+
+const parseInitialState = () => {
+	const storeTag = document.querySelector(
+		`script[type="application/json"]#wc-interactivity-initial-state`
+	);
+	if ( ! storeTag?.textContent ) return {};
+	try {
+		const initialState = JSON.parse( storeTag.textContent );
+		if ( isObject( initialState ) ) return initialState;
+		throw Error( 'Parsed state is not an object' );
+	} catch ( e ) {
+		// eslint-disable-next-line no-console
+		console.log( e );
+	}
+	return {};
+};
+
+export const stores = new Map();
+
+const objToProxy = new WeakMap();
+const proxyToNs = new WeakMap();
+
+const proxify = ( obj: any, ns: string ) => {
+	if ( ! objToProxy.has( obj ) ) {
+		const proxy = new Proxy( obj, handlers );
+		objToProxy.set( obj, proxy );
+		proxyToNs.set( proxy, ns );
+	}
+
+	return objToProxy.get( obj );
+};
+
+const handlers = {
+	get: ( target, key, receiver ) => {
+		const ns = proxyToNs.get( receiver );
+
+		// Check if the proxy is the store root and no prop with that name
+		// exist. In that case, return an empty object for the prop requested.
+		if ( receiver === stores.get( ns ) && ! ( key in target ) ) {
+			const obj = {};
+			target[ key ] = obj;
+			return proxify( obj, ns );
+		}
+
+		// Check if the property is a getter.
+		const getter = Object.getOwnPropertyDescriptor( target, key )?.get;
+		if ( getter ) {
+			const scope = getScope();
+			if ( scope ) {
+				scope.getters = scope.getters || new Map();
+				if ( ! scope.getters.has( getter ) ) {
+					scope.getters.set(
+						getter,
+						computed( () => {
+							setNamespace( ns );
+							setScope( scope );
+							const result = getter.call( target );
+							// resetScope(); // maybe scope should be a stack?
+							resetNamespace();
+							return result;
+						} )
+					);
+				}
+				return scope.getters.get( getter ).value;
+			}
+		}
+
+		const result = Reflect.get( target, key, receiver );
+
+		// Check if the property is a generator.
+		if ( result?.constructor?.name === 'GeneratorFunction' ) {
+			return async ( ...args ) => {
+				const scope = getScope();
+				const gen = result( ...args );
+				const iterate = async ( iteration ) => {
+					resetScope();
+					resetNamespace();
+					if ( iteration.done ) return iteration.value;
+					const res = await iteration.value;
+					setNamespace( proxyToNs.get( receiver ) );
+					setScope( scope );
+					return await iterate( gen.next( res ) );
+				};
+				try {
+					return iterate( gen.next() );
+				} catch ( e ) {
+					resetNamespace();
+					resetScope();
+					throw e;
+				}
+			};
+		}
+
+		// Check if the property is a function.
+		if ( typeof result === 'function' ) {
+			const scope = getScope();
+			return ( ...args ) => {
+				setNamespace( proxyToNs.get( receiver ) );
+				setScope( scope );
+				result( ...args );
+				resetScope();
+				resetNamespace();
+			};
```

Comment:
Oh, I forgot that functions can return 😄 

```suggestion
			return ( ...args ) => {
				setNamespace( proxyToNs.get( receiver ) );
				setScope( scope );
				const ret = result( ...args );
				resetScope();
				resetNamespace();
        return ret;
			};
```

Reply:
Oh, right! Fixed in 95378841448d28206058a944da98180a4a039536.